### PR TITLE
Rework the build system

### PR DESCRIPTION
**Pull request by TAAPArthur**
Added simple script to autodetect if optional dependencies are installed
and enable/disable them as needed. Note this solution uses the compiler
directly instead of an external program like pkgconfig so it doesn't
require any extra dependencies. It is intended to work with any
arbitrary compiler; it has been tested with tcc and gcc.

There are some "breaking" changes hidden here
- HAVE_GIFLIB was renamed to HAVE_LIBGIF to match HAVE_LIBEXIF
- Simply typing `make` will no longer try to build with optional
  dependencies
  
This is helpful for distros like Kiss linux (and probably other source based distros) where the default build dependencies are intended to be minimal. Users could not get the extra optional features by just having those extra dependencies installed without having to modify the default build file.

**Note** that simply running `make && make install` still works. `./configure` is completely optional

### CODE REVIEW

File: configure
Comment (N-R-K) [inline configure:3]:
```suggestion
:> options.mk
```

File: Makefile
Comment (N-R-K) [inline Makefile:13]: If we're going to rename this, I think it's best to rename the `#ifdef`s inside the source as well for consistency.

**Review by N-R-K — CHANGES_REQUESTED** [review]
Since this diverges from the previous build system, I'd wait to hear more opinions (5+ atleast) before merging. I don't have any strong opinion on it.

Also should we add rule for `./configure` inside the Makefile, `make configure` ?

File: configure
Comment (XPhyro) [inline configure:1]:
```suggestion
#!/usr/bin/env sh
```

**Review by XPhyro — CHANGES_REQUESTED** [review]

**Comment by TAAPArthur** [comment]
> Also should we add rule for `./configure` inside the Makefile, `make configure` ?

If we do, would think that make clean would delete the options.mk file. I don't have a strong opinion either way

**Comment by N-R-K** [comment]
> make clean would delete the options.mk file

Sounds good.

- - -

Also, I'd prefer having newer commit instead of force pushes. It makes it easier to see what changed. We can just squash it before merging.

File: Makefile
Comment (N-R-K) [inline Makefile:31]: We should use `LDLIBS` here.

https://www.gnu.org/software/make/manual/html_node/Implicit-Variables.html
>LDFLAGS
> Extra flags to give to compilers when they are supposed to invoke the linker, ‘ld’, such as -L. Libraries (-lfoo) should be added to the LDLIBS variable instead.

File: Makefile
Comment (N-R-K) [inline Makefile:26]: These aren't optimizations, so we should have them set by default.

```suggestion
CFLAGS += -std=c99 -Wall -pedantic
```

**Review by N-R-K — CHANGES_REQUESTED** [review]

File: Makefile
Comment (TAAPArthur) [reply Makefile:26]: They are set by default either way. The question is if they should just be default or extend whatever the use has set. I generally lead towards giving people the ability to change build flags with modifying the makefile. One reason people may want to turn off -Wall is to silence harmless warning messages like ` warning: function might return no value: 'cg_quit'`. Probably not a big deal but just giving the option cause it is free to do so.

Comment (N-R-K) [reply Makefile:26]: > They are set by default either way.

Not if the environment has `CFLAGS` defined.

**Comment by N-R-K** [comment]
Lets rename the PR since it's touching a bit more than just the configure script.

Also a couple other things

* Should we honor environment `PREFIX`  by using `+=` instead ?
* Should we add a var `name` instead of hard-coding `sxiv`/`nsxiv` ?

**Comment by XPhyro** [comment]
> * Should we honor environment `PREFIX`  by using `+=` instead ?
> * Should we add a var `name` instead of hard-coding `sxiv`/`nsxiv` ?

I think
- We should honour the environment `PREFIX`. But I am okay with either.
- We should hardcode the name. I do not see any use in not hardcoding the name.

File: Makefile
Comment (TAAPArthur) [reply Makefile:26]: I see the confusion. By default, I meant if CFLAGS isn't set in the env or directly passed into make.

**Comment by TAAPArthur** [comment]
@XPhyro Why `printf` instead of `echo`. `echo` without flags is just as portable and believe it looks nicer than manually adding `\n` everywhere.

**Comment by N-R-K** [comment]
@XPhyro `printf` is unnecessary here since there's no special formatting. `echo` is also faster since it's a shell built-in.

File: Makefile
Comment (N-R-K) [reply Makefile:26]: Btw users can set `-w` in their CFLAGS to disable warnings.

**Comment by XPhyro** [comment]
`printf` is also a shell built-in. `echo` suffices, but I would rather use `printf` everywhere than use `printf` sometimes and `echo` mostly.

**Comment by N-R-K** [comment]
I don't think it's a good idea to always use something regardless of the context.

File: Makefile
Comment (N-R-K) [reply Makefile:26]: > if CFLAGS isn't set in the env
Yeah, that's what I meant. If it's defined in the env then those flags will get overridden.

**Comment by N-R-K** [comment]
I'd prefer to use `+=` on `CFLAGS` but I'm fine with the current state of this PR either ways. Let's hear some more opinion before proceeding.

**Comment by N-R-K** [comment]
> Also should we add rule for `./configure` inside the Makefile, `make configure` ?

On second thought, let's not do that. `./configure` would need to be run everytime `make clean` is ran. People might be using `config.h` on their clean list and do clean builds. Also I think `./configure` is more widely used, so let's not add any more Makefile rules.

**Review by N-R-K — APPROVED** [review]

**Comment by GRFreire** [comment]
I was just looking to the `options.mk` file and noticed something:
```
HAVE_LIBGIF=1
OPTIONAL_LIBS+=-lgif
```
Aren't those redundant? I mean, if I disable only `HAVE_LIBGIF` it builds without the support of gifs, but if just remove `OPTIONAL_LIBS+=-lgif` it no longer builds. I think that one should only need to change `HAVE_LIBGIF` and the makefile would decide to include the optional dependency or not (kinda the way it was).

Other thing I noticed is that changing the `configure.mk` and running `make && make install` do not work. Maybe add `configure.mk` file to the list of the binary dependencies?

**note:** I am assuming that this file is supposed to be edited by the user, not only generated from the configure script.

**Comment by GRFreire** [comment]
Not sure if that is intentional, but the `make` command is much more verbose (does not look nice to me):
dev-build
configure-build

**Comment by TAAPArthur** [comment]
> Aren't those redundant? 

Yes they are. One annoying thing about this impl is some redudency between configure and Makefil. configure needs to to the -l options to test against the compiler, but so does Makefile if we allow users to manually specify files. I agree the way it is isn't ideal. Will modify to have configure just set the HAVE_X and have a series of `ifeq` to set the flag.

> Other thing I noticed is that changing the `configure.mk` and running `make && make install` do not work. Maybe add `configure.mk` file to the list of the binary dependencies?

I also run make clean whenever I re-run configure. Feel like that is standard, but can't point to any source to back that up off hand

> **note:** I am assuming that this file is supposed to be edited by the user, not only generated from the configure script.

My intent was it to not be edited by the user. If the user wants their own options they can just pass them to make directly.

>Not sure if that is intentional, but the make command is much more verbose now (does not look nice to me):

I changed the behavior when debugging. Personally I prefer makefile to be verbose so I can see what went wrong. It is kinda tangential to the overall goal of the PR though and apparently opinionated so won't complain if it gets reverted

**Comment by GRFreire** [comment]
> I changed the behavior when debugging. Personally I prefer makefile to be verbose so I can see what went wrong. It is kinda tangential to the overall goal of the PR though and apparently opinionated so won't complain if it gets reverted

I prefer the way it once was, has cleaner output and when something does go wrong, it outputs the error.
error-build

Although this is a rework of the build system, I dont feel like changing this debug thing. Maybe with some flag idk, just not the default.

**Comment by N-R-K** [comment]
User can run `make -s` to make it silent. I personally like the more verbose output, but I'm fine with any.

- - -

> I am assuming that this file is supposed to be edited by the user, not only generated from the configure script.

One more option worth considering might be to drop the script entirely and go for `config.mk` similar to other suckless projects and have the user edit it manually as they wish.

**Comment by N-R-K** [comment]
Kept `version` lower letters, since the preprocessor directive is all-caps.

**Comment by TAAPArthur** [comment]
> One more option worth considering might be to drop the script entirely and go for `config.mk` similar to other suckless projects and have the user edit it manually as they wish.

@N-R-K so following @GRFreire's suggestion, users can create and manually modify `options.mk` if they so chose. I'm not particular about the name. `./configure` will overwrite the the makefile but if that isn't terrible we could rename it to `config.mk`.

File: configure
Comment (kdkasad) [reply configure]: Just wondering: why change this?

**Comment by kdkasad** [comment]
Personally, I don't like this set of changes. If the goal is autodetecting the optional dependencies, why not just do that from the Makefile itself? Having a separate shell script just makes the build much more complicated as you now have to keep track of two different build scripts (`configure` and `Makefile`) as well as the `options.mk` file.

I suggest doing any of the following:
1. Just have the options in the `Makefile` or in a `config.mk` file and don't do any autodetecting. This is what many suckless-style programs do.
2. Autodetect the optional dependencies in the Makefile if they are not set by the user on the command line.
3. Combine both (1) and (2); have a config.mk file, but instead of just `HAVE_LIBxxx=1` or `HAVE_LIBxxx=0`, include a third option for `HAVE_LIBxxx=auto`, in which case some code in the Makefile will autodetect the dependency library.

**Comment by N-R-K** [comment]
I'm fine with 1. Let's not do any autodetecting at all, and unload some of the configuration variables into `config.mk`.

**Comment by TAAPArthur** [comment]
>  Having a separate shell script just makes the build much more complicated as you now have to keep track of two different build scripts (`configure` and `Makefile`) as well as the `options.mk` file.

@kdkasad Really? The configure script is completely optional so from a end-user perspective it is harmless if you don't want it. From a maintenance view, adding a new optional depends involves modifying 1 extra line in the config file compared to it not being there.

**Comment by N-R-K** [comment]
No lines need to be modified though. Afaik K1ss builds are just shell scripts, in which case `make HAVE_LIBGIF=0` would suffice.

**Comment by kdkasad** [comment]
> @kdkasad Really? The configure script is completely optional so from a end-user perspective it is harmless if you don't want it.

Right, but I'm talking about if you do want it. If I want autodetection, I have to run the configure script. If I'm a new user, it could be confusing. E.g.:
* If I `make clean`, do I need to re-configure?
* Do I need to edit `options.mk` like I do `config.mk` in other programs?
* Do I need to `./configure` at all?
* How to I switch from autodetection to manual configuration?
* Normally things like install prefix (i.e. `$(PREFIX)`)  are handled by `./configure`, but for some reason that doesn't work here?

While the answers seem obvious to the designer(s) of the build system, they could be difficult for new users.

In contrast, the three options I proposed make the answers to those questions obvious (or so I think. Again, the above statement applies to me, too).

**Comment by N-R-K** [comment]
Changed everything to be more suckless style, closely mimicking the `dwm` build system.

Pushed it purely for review purposes, @TAAPArthur @kdkasad please provide some feedback. And feel free to revert if needed.

**Comment by TAAPArthur** [comment]
I see. I wasn't thinking of the new user. I personally don't think the answers to your questions are hard to figure out, but I agree that everything is obvious when you know the answer.

>Another note: There are two well-known conventions: (1) no configure script, just a Makefile, or (2) a configure script which does all the configuration, generates the Makefile, etc. The changes proposed in this PR follow neither convention. My main point is that we should at least follow one.

Fair point. If the concern is that the configure script itself is potentially confusing, that's easy enough to deal with. Could add something like
```
config.mk:
   ./configure # Or just inline the shell code
```
Maybe even make configure a default target so config.mk will be generated by invoking make if it doesn't exist.
Which may have been what you meant with option 2. I had interpreted it as preferring make-script over shell rather than looking at it from a UI perspective.

@N-R-K I don't think I'm a fine of your change since the original goal of this PR was auto-detection. I think I understand @kdkasad's concern now and have a more convenient solution in mind

**Comment by TAAPArthur** [comment]
Ok so removed `configure`
Now if you run `make`, config.mk will be auto-generated if it doesn't exist and will enable all optional depends possible. Users can still pass in whatever flags they want to make and those will take precedence. So basically the default behavior has just been changed to build with all optional dependencies possible. I think this is a sane default.

**Comment by N-R-K** [comment]
I'm mostly fine with this. But lets have the configure script be silent. Another thing is if `config.mk` doesn't exist and `make clean` is run, config script gets run too, I don't like this.

Finally should we go back to having a more silent output like @GRFreire suggested?

**Comment by N-R-K** [comment]
@TAAPArthur One more thing

```sh
# enable features requiring giflib (-lgif)
HAVE_LIBGIF ?= 0
# enable features requiring libexif (-lexif)
HAVE_LIBEXIF ?= 0
```

Let's remove this from the makefile. Editing it here doesn't really work since config.mk takes priority, this can be confusing for new users trying to remove a dep. Proper way to remove it should be either config.mk or through passing option to make itself.

**Comment by TAAPArthur** [comment]
> Let's remove this from the makefile. Editing it here doesn't really work since config.mk takes priority, this can be confusing for new users trying to remove a dep. Proper way to remove it should be either config.mk or through passing option to make itself.

So if we remove those lines and the user removes the config.mk for whatever reason, the program won't build. If you remove the `?` the variables will override those in config.mk. I don't really disagree about the recommended way to change vars, but think it would be best it config.mk could be empty and the program still builds.

About the `make clean`... good catch.
Could do something like
```
ifneq (clean,$(filter clean,$(MAKECMDGOALS)))
-include config.mk
endif
```
To not get config.mk built when running clean. Perhaps should extend to uninstall too

**Comment by N-R-K** [comment]
> So if we remove those lines and the user removes the config.mk for whatever reason, the program won't build.

Removing `config.mk` will just regen it. If by remove you meant editing it, then I think I have a solution for that.

**Comment by N-R-K** [comment]
I think we're mostly done here now. Just need to decide do we go back to silent output or keep it verbose.

**Review by N-R-K — APPROVED** [review]

**Comment by GRFreire** [comment]
> I think we're mostly done here now. Just need to decide do we go back to silent output or keep it verbose.

You know my opinion about this, but since apparently most of y'all prefer the verbose output, we should go with it. 

> Also if we go verbose, then lets remove some of the `echo`s since they no longer have any purpose.

It does have a purpose. Lets say I run `make -s`, I don't want all the commands that it is running, but still want to know some idea if something is happening (compiling again, just installing, clean, etc).

**Comment by TAAPArthur** [comment]
> You know my opinion about this, but since apparently most of y'all prefer the verbose output, we should go with it.

Out of curiosity, assuming the minimal output isn't specific to nsix, would setting `export MAKEFLAGS=-s` in your profile resolve the inconvenience for you? It should cause make to be invoked with `-s` unless explicitly overridden with `--no-silent`

**Comment by N-R-K** [comment]
> but still want to know some idea if something is happening (compiling again, just installing, clean, etc).

Install and uninstall still has the `echo`s btw.

**Comment by kdkasad** [comment]
I also prefer the old, less verbose style. However, I understand the benefit of the verbose version.

As far as I can tell it's easier to disable the printing of commands in verbose version (you just use `-s`) than it is to enable them when in the silent version. Therefore, we should go with the verbose mode.

**Comment by GRFreire** [comment]
> Install and uninstall still has the echos btw.

Really? When I run `make -s` it has no output at all

File: README.md
Comment (GRFreire) [inline README.md:69]: Lets say you have already compiled (without the flag), it will not recompile if you pass `HAVE_LIBGIF=0`, instead it will print `make: Nothing to be done for 'all'`. You would need to run `make clean` explicitly.
Idk if the makefile should check for this, but since you are running make with a new flag, I feel like it should recompile.

Comment (N-R-K) [reply README.md:69]: This might be an inherent problem with how make works. It doesn't really track the variables, only the dependencies.

If we can trigger a rebuild on new flags that would be ideal. But I'm not sure if it's possible.

Comment (GRFreire) [reply README.md:69]: I suspected that it wasnt possible, but wanted to point out to see if others have any idea (or even say that its expected behavior).

Comment (N-R-K) [reply README.md:69]: I wouldn't say its _expected_ but rather a limitation of `make`. @TAAPArthur seems to know more about make than I do, maybe he has some trick up his sleeve for this?

Comment (kdkasad) [reply README.md:69]: The only thing I can think of is writing the variables to a file and checking the difference every time `make` is run. But I think that's too hacky and we should just advise running `make clean` first.

**Comment by N-R-K** [comment]
@GRFreire Can you run `make -s` and confirm that everything works as expected?

I'm fine with the echos, and I think taap didn't oppose it either. I'd like to move forward without disappointing anyone.

**Review by GRFreire — APPROVED** [review]

**Comment by GRFreire** [comment]
> Can you run `make -s` and confirm that everything works as expected?

Everything working as expected.

File: configure
Comment (N-R-K) [reply configure]: Merge is being blocked because of this. @XPhyro can you approve (assuming you're fine with the PR)?

**Review by kdkasad — APPROVED** [review]

**Pull request merged**
